### PR TITLE
Node queries

### PR DESCRIPTION
**Pull request by EmilLa1**
## Purpose

- NodeInfo
- PeersInfo
- Network dumps
- Banning
- Peer connect/disconnect
- Shutdown

## Changes

Introduced types and endpoints.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-node/src/grpc2.rs [hunk 1]
```diff
@@ -264,7 +265,10 @@ pub mod server {
     use std::sync::{Arc, Mutex};
     use tonic::{async_trait, transport::ServerTlsConfig};
 
-    use super::*;
+    use super::{
+        types::node_info::{self, baker_consensus_info},
```
Comment (abizjak) [inline concordium-node/src/grpc2.rs:269]: Probably just qualify these. That's what we do with others.

File: concordium-node/src/grpc2.rs [hunk 2]
```diff
@@ -985,6 +988,132 @@ pub mod server {
...
             Ok(response)
         }
+
+        async fn get_node_info(
+            &self,
+            _request: tonic::Request<crate::grpc2::types::Empty>,
+        ) -> Result<tonic::Response<types::NodeInfo>, tonic::Status> {
+            let peer_version = self.node.get_version();
+            let node_type = match self.node.peer_type() {
+                crate::common::PeerType::Bootstrapper => node_info::NodeType::Bootstrapper as i32,
```
Comment (abizjak) [inline concordium-node/src/grpc2.rs:998]: Isn't there an `into` instead?  `as` is  not great for maintainability.

File: concordium-node/src/grpc2.rs [hunk 3]
```diff
@@ -1367,6 +1370,374 @@ pub mod server {
...
+                Err(tonic::Status::internal("Could not load banned peers."))
+            }
+        }
+
+        async fn ban_peer(
+            &self,
+            request: tonic::Request<crate::grpc2::types::PeerToBan>,
+        ) -> Result<tonic::Response<crate::grpc2::types::Empty>, tonic::Status> {
+            let ip = request.into_inner().ip_address.require()?;
+            match <std::net::IpAddr as std::str::FromStr>::from_str(&ip.value) {
```
Comment (abizjak) [inline concordium-node/src/grpc2.rs:1465]: What is the reason for this rather verbose style? Why can't we use

```
ip.value.parse::<std::net::IpAddr>()
```

File: concordium-node/src/grpc2.rs [hunk 4]
```diff
@@ -1367,6 +1370,374 @@ pub mod server {
...
+                    e
+                ))),
+            }
+        }
+
+        async fn unban_peer(
+            &self,
+            request: tonic::Request<crate::grpc2::types::BannedPeer>,
+        ) -> Result<tonic::Response<crate::grpc2::types::Empty>, tonic::Status> {
+            match <std::net::IpAddr as std::str::FromStr>::from_str(
```
Comment (abizjak) [inline concordium-node/src/grpc2.rs:1481]: Likewise here

File: concordium-node/src/grpc2.rs [hunk 5]
```diff
@@ -1367,6 +1370,374 @@ pub mod server {
...
+            }
+        }
+
+        #[cfg(feature = "network_dump")]
+        async fn dump_start(
+            &self,
+            request: tonic::Request<crate::grpc2::types::DumpRequest>,
+        ) -> Result<tonic::Response<crate::grpc2::types::Empty>, tonic::Status> {
+            let file_path = request.get_ref().file.to_owned();
+            // todo: fail with invalid argument if file path was empty.
```
Comment (abizjak) [inline concordium-node/src/grpc2.rs:1505]: TODO, please handle as described.

File: concordium-node/src/grpc2.rs [hunk 6]
```diff
@@ -1367,6 +1370,374 @@ pub mod server {
...
+                .activate_dump(
+                    if file_path.is_empty() {
+                        "dump"
+                    } else {
+                        &file_path
+                    },
+                    request.get_ref().raw,
+                )
+                .is_ok();
+            Ok(tonic::Response::new(crate::grpc2::types::Empty {}))
```
Comment (abizjak) [inline concordium-node/src/grpc2.rs:1517]: Would it make sense to be more defensive here and only allow relative paths inside, e.g., a data directory?

**Review by abizjak — APPROVED** [review]
This looks good to me, modulo the changes that are going to be needed as a result of the changes in the types.proto definition.

File: concordium-node/src/grpc2.rs (hunk 6 above)
Comment (EmilLa1) [reply concordium-node/src/grpc2.rs]: I'm not sure. We assume the operator has exclusive access to his node anyhow - why not let him dump the file where ever he wants to? 
On the other hand it also sounds fine to somehow sanitize the path. 

I'll let that decision be up to the reviewer :) ...

Comment (abizjak) [reply concordium-node/src/grpc2.rs]: I guess as long as it doesn't crash the node it should be fine. So let's keep it simple.

**Pull request merged**